File: HamamatsuCamera/dcamsdk4/samples/vbnet/vbAcq4/subacq4.vb
' subacq4.vb

' implementation of subacq.copydib() function
' instead of using DLL, this source code requires /unsafe option.

Imports System.Runtime.InteropServices
Imports System.Drawing.Imaging
Imports vbAcq4.Hamamatsu.DCAM4

#If Not USE_SUBACQ_DLL Then

Namespace Hamamatsu
    Public Class subacq4

        Public Enum SUBACQERR
            ' error
            NO_DLL = &H80000000
            INVALID_ARG = &H80000001
            INVALID_DST = &H80000011
            INVALID_DSTPIXELFORMAT = &H80000012
            INVALID_SRC = &H80000021
            INVALID_SRCPIXELTYPE = &H80000022

            NOSUPPORT_LUTARRAY = &H80000101

            ' success
            SUCCESS = 1
        End Enum

        Public Class subacq

            Public Declare Sub CopyMemory_UShortarrayFromIntPtr Lib "kernel32" Alias "RtlMoveMemory" (ByVal destination() As UShort, ByVal Source As IntPtr, ByVal Length As Integer)

            Public Shared Function copydib(ByRef bitmap As Bitmap, ByVal src As DCAMBUF_FRAME, ByRef rect As Rectangle, ByVal lutmax As Integer, ByVal lutmin As Integer) As SUBACQERR
                Dim w As Integer
                Dim h As Integer

                w = rect.Width
                h = rect.Height
                If w > bitmap.Width Then w = bitmap.Width
                If h > bitmap.Height Then h = bitmap.Height
                If w > src.width Then w = src.width
                If h > src.height Then h = src.height

                Dim dst As BitmapData
                dst = bitmap.LockBits(rect, ImageLockMode.ReadWrite, PixelFormat.Format32bppRgb)

                Dim err As SUBACQERR
                If src.type = DCAM_PIXELTYPE.MONO16 Then
                    err = copydib_rgb32_from_mono16(dst.Scan0, dst.Stride, src.buf, src.rowbytes, w, h, lutmax, lutmin)
                Else
                    err = SUBACQERR.INVALID_SRCPIXELTYPE
                End If

                bitmap.UnlockBits(dst)
                Return err
            End Function

            Private Shared Function copydib_rgb32_from_mono16(ByVal dst As IntPtr, ByVal dstrowbytes As Integer, ByVal src As IntPtr, ByVal srcrowbytes As Int32, ByVal width As Integer, ByVal height As Integer, ByVal lutmax As Integer, ByVal lutmin As Integer) As SUBACQERR
                Dim s As UShort()
                Dim d As Int32()
                s = New UShort(width) {}
                d = New Int32(width) {}

                Dim gain As Double
                Dim inBase As Double
                gain = 0
                inBase = 0

                If lutmax <> lutmin Then
                    If lutmin < lutmax Then
                        gain = 256.0 / (lutmax - lutmin + 1)
                        inBase = lutmin
                    ElseIf lutmin > lutmax Then
                        gain = 256.0 / (lutmax - lutmin - 1)
                        inBase = lutmax
                    End If
                ElseIf lutmin > 0 Then  ' binary threshold
                    gain = 0
                    inBase = lutmin
                End If

                Dim y As Integer
                For y = 0 To height - 1
                    Dim offset As Integer

                    offset = srcrowbytes * y
                    CopyMemory_UShortarrayFromIntPtr(s, src.ToInt64() + offset, srcrowbytes)

                    copydibline_rgb32_from_mono16(d, s, width, gain, inBase)

                    offset = dstrowbytes * y
                    Marshal.Copy(d, 0, dst.ToInt64() + offset, width)
                Next
                Return SUBACQERR.SUCCESS
            End Function

            Private Shared Sub copydibline_rgb32_from_mono16(ByVal d() As Int32, ByVal s() As UShort, ByVal width As Integer, ByVal gain As Double, ByVal inBase As Double)
                Dim x As Integer
                If gain <> 0 Then
                    For x = 0 To width - 1
                        Dim u As UInt16
                        u = s(x)

                        Dim v As Double
                        v = gain * (u - inBase)

                        Dim c As Byte
                        If v > 255 Then
                            c = 255
                        ElseIf v < 0 Then
                            c = 0
                        Else
                            c = v
                        End If

                        d(x) = Color.FromArgb(c, c, c).ToArgb()
                    Next
                ElseIf inBase > 0 Then
                    For x = 0 To width - 1
                        Dim u As UInt16
                        u = s(x)

                        Dim c As Byte
                        If u >= inBase Then
                            c = 255
                        Else
                            c = 0
                        End If

                        d(x) = Color.FromArgb(c, c, c).ToArgb()
                    Next
                Else
                    For x = 0 To width - 1
                        Dim u As UInt16
                        u = s(x)

                        Dim c As Byte
                        c = (u >> 8)

                        d(x) = Color.FromArgb(c, c, c).ToArgb()
                    Next
                End If
            End Sub
        End Class
    End Class
End Namespace

#End If ' Not USE_SUBACQ_DLL

File: HamamatsuCamera/dcamsdk4/samples/cs/csAcq4/subacq4.cs
// subacq4.cs

// implementation of subacq.copydib() function. SUBACQ4.DLL is not necessary.

using System;
using System.Runtime.InteropServices;
using System.Drawing;
using System.Drawing.Imaging;
using Hamamatsu.DCAM4;

#if ! USE_SUBACQ_DLL

namespace Hamamatsu.subacq4
{
    public enum SUBACQERR : uint
    {
        // error
        NO_DLL = 0x80000000,
        INVALID_ARG = 0x80000001,
        INVALID_DST = 0x80000011,
        INVALID_DSTPIXELFORMAT = 0x80000012,
        INVALID_SRC = 0x80000021,
        INVALID_SRCPIXELTYPE = 0x80000022,

        NOSUPPORT_LUTARRAY = 0x80000101,

        // success
        SUCCESS = 1
    }

    public class subacq
    {
        public static SUBACQERR copydib(ref Bitmap bitmap, DCAMBUF_FRAME src, ref Rectangle rect, int lutmax, int lutmin, int bpp)
        {
            int w = rect.Width;
            int h = rect.Height;
            if (w > bitmap.Width) w = bitmap.Width;
            if (h > bitmap.Height) h = bitmap.Height;
            if (w > src.width) w = src.width;
            if (h > src.height) h = src.height;

            SUBACQERR err;
            BitmapData dst = bitmap.LockBits(rect, ImageLockMode.ReadWrite, PixelFormat.Format24bppRgb);

            if (src.type == DCAM_PIXELTYPE.MONO16)
                err = copydib_rgb24_from_mono16(dst.Scan0, dst.Stride, src.buf, src.rowbytes, w, h, lutmax, lutmin, bpp);
            else
                err = SUBACQERR.INVALID_SRCPIXELTYPE;

            bitmap.UnlockBits(dst);

            return err;
        }

        //public static SUBACQERR copydib(ref Bitmap bitmap, DCAMBUF_FRAME src, ref Rectangle rect, int lutmax, int lutmin, int bpp)
        //{
        //    int w = rect.Width;
        //    int h = rect.Height;
        //    if (w > bitmap.Width) w = bitmap.Width;
        //    if (h > bitmap.Height) h = bitmap.Height;
        //    if (w > src.width) w = src.width;
        //    if (h > src.height) h = src.height;

        //    SUBACQERR err;
        //    BitmapData dst = bitmap.LockBits(rect, ImageLockMode.ReadWrite, PixelFormat.Format8bppIndexed);

        //    if (src.type == DCAM_PIXELTYPE.MONO16)
        //        err = copydib_gray8_from_mono16(dst.Scan0, dst.Stride, src.buf, src.rowbytes, w, h, lutmax, lutmin, bpp);
        //    else
        //        err = SUBACQERR.INVALID_SRCPIXELTYPE;

        //    bitmap.UnlockBits(dst);

        //    // Apply grayscale palette (required for 8-bit images)
        //    ColorPalette pal = bitmap.Palette;
        //    for (int i = 0; i < 256; i++)
        //        pal.Entries[i] = Color.FromArgb(i, i, i);
        //    bitmap.Palette = pal;

        //    return err;
        //}

    //    private static SUBACQERR copydib_gray8_from_mono16(
    //IntPtr dst, Int32 dstrowbytes, IntPtr src, Int32 srcrowbytes, Int32 width, Int32 height,
    //Int32 lutmax, Int32 lutmin, Int32 bpp)
    //    {
    //        Int16[] s = new Int16[width];
    //        byte[] d = new byte[dstrowbytes];

    //        double gain = 256.0 / (lutmax - lutmin + 1);
    //        double inBase = lutmin;

    //        for (int y = 0; y < height; y++)
    //        {
    //            Int32 offsetSrc = srcrowbytes * y;
    //            Marshal.Copy((IntPtr)(src.ToInt64() + offsetSrc), s, 0, width);

    //            for (int x = 0; x < width; x++)
    //            {
    //                UInt16 u = (UInt16)s[x];
    //                double v = gain * (u - inBase);

    //                d[x] = (Byte)(v > 255 ? 255 : (v < 0 ? 0 : v)); // Clamp to 0-255
    //            }

    //            Int32 offsetDst = dstrowbytes * y;
    //            Marshal.Copy(d, 0, (IntPtr)(dst.ToInt64() + offsetDst), width);
    //        }
    //        return SUBACQERR.SUCCESS;
    //    }

        private static SUBACQERR copydib_rgb24_from_mono16(IntPtr dst, Int32 dstrowbytes, IntPtr src, Int32 srcrowbytes, Int32 width, Int32 height, Int32 lutmax, Int32 lutmin, Int32 bpp)
        {
            Int16[] s = new Int16[width];
            byte[] d = new byte[dstrowbytes];

            double gain = 0;
            double inBase = 0;

            if (lutmax != lutmin)
            {
                if (lutmin < lutmax)
                {
                    gain = 256.0 / (lutmax - lutmin + 1);
                    inBase = lutmin;
                }
                else
                if (lutmin > lutmax)
                {
                    gain = 256.0 / (lutmax - lutmin - 1);
                    inBase = lutmax;
                }
            }
            else
            if (lutmin > 0)    // binary threshold
            {
                gain = 0;
                inBase = lutmin;
            }

            Int16 y;
            for (y = 0; y < height; y++)
            {
                Int32 offset;

                offset = srcrowbytes * y;
                Marshal.Copy((IntPtr)(src.ToInt64() + offset), s, 0, width);

                copydibline_rgb24_from_mono16(d, s, width, gain, inBase, bpp);

                offset = dstrowbytes * y;
                Marshal.Copy(d, 0, (IntPtr)(dst.ToInt64() + offset), dstrowbytes);
            }
            return SUBACQERR.SUCCESS;
        }

        private static void copydibline_rgb24_from_mono16(byte[] d, Int16[] s, Int32 width, double gain, double inBase, Int32 bpp)
        {
            Int16 x;
            Int16 i = 0;
            if (gain != 0)
            {
                for (x = 0; x < width; x++)
                {
                    UInt16 u = (UInt16)s[x];

                    double v = gain * (u - inBase);

                    Byte c;
                    if (v > 255)
                        c = 255;
                    else
                    if (v < 0)
                        c = 0;
                    else
                        c = (Byte)v;

                    d[i++] = c;
                    d[i++] = c;
                    d[i++] = c;
                }
            }
            else
            if (inBase > 0)    // binary threshold
            {
                for (x = 0; x < width; x++)
                {
                    UInt16 u = (UInt16)s[x];

                    Byte c = (Byte)(u >= inBase ? 255 : 0);

                    d[i++] = c;
                    d[i++] = c;
                    d[i++] = c;
                }
            }
            else
            {
                for (x = 0; x < width; x++)
                {
                    UInt16 u = (UInt16)s[x];

                    Byte c = (Byte)(u >> (bpp - 8));

                    d[i++] = c;
                    d[i++] = c;
                    d[i++] = c;
                }
            }
        }
    }
}

#endif // ! USE_SUBACQ_DLL
